src/service/AddressService.java

```diff
@@ -0,0 +1,163 @@
+
+            while(rs.next()) {
+                houseNum = rs.getInt("house_num");
+                street = rs.getString("street");
+                postalCode = rs.getString("postal_code");
+                city = rs.getString("city");
+                province = rs.getString("province");
+                cAddress = new Address(houseNum, street, postalCode, city, province);
+                customerAddresses.add(cAddress);
+            }
+            pstmt.close();
+            return customerAddresses;
+
+        } catch (SQLException ex) {
+            System.out.println("Message: " + ex.getMessage());
+        }
+
+        return null;
+    }
+
+    // add a new address for a given customer
+    public void addAddress(String username, int houseNum, String street, String postalCode,
+                           String city, String province) {
+        PreparedStatement pstmt;
+
+        try {
+
+            // Insert into Address table if not already there
+            pstmt = connection.prepareStatement(
+                    "INSERT INTO Address VALUES (?, ?, ?)"
+            );
+            pstmt.setInt(1, houseNum);
+            pstmt.setString(2, street);
+            pstmt.setString(3, postalCode);
+            pstmt.executeUpdate();
+            pstmt.close();
+
+            // Insert into City_Province if not already there
+            pstmt = connection.prepareStatement(
+                    "INSERT INTO City_Province VALUES (?, ?, ?)"
+            );
+            pstmt.setString(1, city);
+            pstmt.setString(2, province);
+            pstmt.setString(3, postalCode);
+            pstmt.executeUpdate();
+            pstmt.close();
+
+            // Insert into Customer_Has_Address table
+            pstmt = connection.prepareStatement(
+                    "INSERT INTO Customer_Has_Address VALUES (?, ?, ?, ?)"
+            );
+            pstmt.setString(1, username);
+            pstmt.setInt(2, houseNum);
+            pstmt.setString(3, street);
+            pstmt.setString(4, postalCode);
+            pstmt.executeUpdate();
+            pstmt.close();
+
+        } catch (SQLException ex) {
+            System.out.println("Message: " + ex.getMessage());
+        }
+    }
+
+    // delete an address
+    public void deleteAddress(String username, int houseNum, String street, String postalCode) {
+        try {
+            PreparedStatement pstmt = connection.prepareStatement(
+                    "DELETE FROM Customer_Has_Address WHERE (username = ?) and (house_num = ?) " +
+                            "and (street = ?) and (postal_code = ?)"
+            );
+            pstmt.setString(1, username);
+            pstmt.setInt(2, houseNum);
+            pstmt.setString(3, street);
+            pstmt.setString(4, postalCode);
+            pstmt.executeUpdate();
+            pstmt.close();
+
+        } catch(SQLException ex) {
+            System.out.println("Message: " + ex.getMessage());
```

Comment:
If you look at what I've done with the SubscriptionService class, I actually try to do a rollback if a SQLException is caught. If the rollback fails, then I throw an exception which I was thinking should be caught in the client side. (I noticed this approach in the branch.java). When this exception is caught, we create an error dialog with an appropriate error message. What are your thoughts?